geom/alg_point_in_ring.go

```diff
@@ -42,25 +46,63 @@ func pointRingSide(pt XY, ring LineString) side {
 		if !ok {
 			continue
 		}
-		inter := ln.intersectLine(ray)
-		if inter.empty {
-			continue
+		if incrementCountPointInRing(pt, ray, ln) {
+			count++
 		}
-		if inter.ptA != inter.ptB {
-			continue
+	}
+	if count%2 == 1 {
+		return interior
+	}
+	return exterior
+}
+
+func incrementCountPointInRing(pt XY, ray, iterLine line) bool {
+	inter := ray.intersectLine(iterLine)
+	if inter.empty {
+		return false
+	}
+	if inter.ptA != inter.ptB {
+		return false
+	}
+	if inter.ptA == iterLine.a || inter.ptA == iterLine.b {
+		otherY := iterLine.a.Y
+		if inter.ptA == iterLine.a {
+			otherY = iterLine.b.Y
 		}
-		if inter.ptA == ln.a || inter.ptA == ln.b {
-			otherY := ln.a.Y
-			if inter.ptA == ln.a {
-				otherY = ln.b.Y
-			}
-			if otherY < pt.Y {
-				count++
-			}
-		} else {
-			count++
+		return otherY < pt.Y
+	}
+	return true
+}
+
+func relatePointToPolygon(pt XY, polyBoundary indexedLines) side {
+	var onBoundary bool
+	ptBox := rtree.Box{MinX: pt.X, MinY: pt.Y, MaxX: pt.X, MaxY: pt.Y}
+	polyBoundary.tree.Search(ptBox, func(i int) error {
+		ln := polyBoundary.lines[i]
+		if ln.intersectsXY(pt) {
+			onBoundary = true
+			return rtree.Stop
 		}
+		return nil
+	})
+	if onBoundary {
+		return boundary
+	}
+
+	extent, ok := polyBoundary.tree.Extent()
+	if !ok {
+		return exterior
 	}
+	ray := line{pt, XY{extent.MaxX + 1, pt.Y}}
```

Comment:
We could be a bit smarter about how we construct the ray. We should construct the ray such that it is the shortest possible distance from the `pt` to the outside of `extent`. This could be as simple as checking the lengths of the rays constructed by going from the pt to the outside of the envelope in the 4 cardinal directions. I'll create an additional ticket for this.